lookout/style/format/annotations/annotations.py

```diff
@@ -0,0 +1,196 @@
+
+
+class Annotation:
+    """Base class for annotation."""
+
+    def __init__(self, start: int, stop: int):
+        """
+        Initialization.
+
+        :param start: Annotation start.
+        :param stop: Annotation end.
+        """
+        self._range = (start, stop)
+        self._start = start
+        self._stop = stop
+
+    start = property(lambda self: self._start)
+
+    stop = property(lambda self: self._stop)
+
+    range = property(lambda self: self._range)
+
+    name = property(lambda self: type(self).__name__)
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        return "%s[%d, %d)" % (self.name, self.start, self.stop)
+
+
+class LineAnnotation(Annotation):
+    """Line number annotation."""
+
+    def __init__(self, start: int, stop: int, number: int):
+        """Init."""
+        super().__init__(start, stop)
+        self._number = number
+
+    number = property(lambda self: self._number)
+
+
+class UASTNodeAnnotation(Annotation):
+    """UAST Node annotation."""
+
+    def __init__(self, start: int, stop: int, node: "bblfsh.Node"):
+        """Init."""
+        super().__init__(start, stop)
+        self._node = node
+
+    node = property(lambda self: self._node)
+
+    @staticmethod
+    def from_node(node: "bblfsh.Node") -> "UASTNodeAnnotation":
+        """Create the annotation from bblfsh node."""
+        return UASTNodeAnnotation(node.start_position.offset, node.end_position.offset, node)
+
+
+class UASTAnnotation(UASTNodeAnnotation):
+    """Full UAST of the file annotation."""
+
+    def __init__(self, start: int, stop: int, uast: "bblfsh.Node"):
+        """
+        Create new UASTAnnotation instance.
+
+        Parents mapping for the tree is built during init.
+
+        :param start: Annotation start.
+        :param stop: Annotation end.
+        :param uast: UAST.
+        """
+        super().__init__(start, stop, uast)
+        self._parents = get_uast_parents(uast)
+
+    uast = property(lambda self: self._node)
+
+    parents = property(lambda self: self._parents)
+
+
+class TokenAnnotation(Annotation):
+    """Annotation for virtual token in the сode."""
+
+    def __init__(self, start: int, stop: int,
+                 uast_annotation: Optional[UASTNodeAnnotation] = None):
+        """
+        Initialization.
+
+        :param start: Annotation start.
+        :param stop: Annotation end.
+        :param uast_annotation: Related UASTNodeAnnotation Annotation if applicable.
+        """
+        super().__init__(start, stop)
+        self._uast_annotation = uast_annotation
+
+    uast_annotation = property(lambda self: self._uast_annotation)
+
+    @property
+    def node(self) -> "bblfsh.Node":
+        """
+        Get UAST Node from related UASTNodeAnnotation.
+
+        :return: related bblfsh UAST Node. None if there is no related annotation.
+        """
+        return self._uast_annotation.node if self._uast_annotation else None
```

Comment:
So we literally annotate with another annotation. Why not carry the node directly?

Reply:
Well, there is an underlying discussion with @m09 https://github.com/src-d/style-analyzer/pull/654#issuecomment-469636019 
@m09, can you give me help to answer this?

Reply:
For me it's better to link to the annotation (not the bblfsh node). The goal of UASTNodeAnnotations being to enhance bblfsh.Node for our use case (here with parents for example, and potentially role/internalrole mapping, etc).